src/AmqpBundle/Amqp/Producer.php

```diff
@@ -51,14 +52,25 @@ public function publishMessage($message, $flags = AMQP_NOPARAM, array $attribute
                       array_merge($this->exchangeOptions['publish_attributes'], $attributes));
 
         $routingKeys = !empty($routingKeys) ? $routingKeys : $this->exchangeOptions['routing_keys'];
-        if (!$routingKeys) {
-            return $this->call($this->exchange, 'publish', [$message, null, $flags, $attributes]);
+
+        $prePublishEvent = new PrePublishEvent($message, $routingKeys, $flags, $attributes);
+
+        if ($this->eventDispatcher) {
+            $this->eventDispatcher->dispatch(PrePublishEvent::NAME, $prePublishEvent);
+
+            if (!$prePublishEvent->canPublish()) {
+                return true;
+            }
+        }
+
+        if (!$prePublishEvent->getRoutingKeys()) {
```

Comment:
I think it's always better to store the response of a method (even if it's a getter... it can be heavier than we thought), because it duplicates the code anyway.